pkg/graveler/graveler.go

```diff
@@ -862,15 +875,127 @@ func generateStagingToken(repositoryID RepositoryID, branchID BranchID) StagingT
 }
 
 func (g *KVGraveler) CreateBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID, ref Ref) (*Branch, error) {
-	return g.db.CreateBranch(ctx, repositoryID, branchID, ref)
+	repo, err := g.RefManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return nil, fmt.Errorf("get repository: %w", err)
+	}
+
+	reference, err := g.Dereference(ctx, repositoryID, ref)
+	if err != nil {
+		return nil, fmt.Errorf("source reference '%s': %w", ref, err)
+	}
+	if reference.ResolvedBranchModifier == ResolvedBranchModifierStaging {
+		return nil, fmt.Errorf("source reference '%s': %w", ref, ErrCreateBranchNoCommit)
+	}
+
+	_, err = g.RefManager.GetBranch(ctx, repositoryID, branchID)
```

Comment:
`GetBranch` hides a call to `GetRepository`, which should already be covered by the call to `GetRepository` at the beginning of the function. In addition, `CreateBrnach` below also calls `GetRepository` and I'm not sure how `Dereference` above works, but it might add another call (?)
I think we should implement kind of `getBranchInternal`, `createBranchInternal` etc, that receive, in addition to the current respective functions parameters, the `Repository` object retrieved by the first (and ideally only) call to `GetRepository`

Reply:
Dereference does not call GetRepository but essentially you are right. Unfortunately this is temp constriction as we don't want to bloat the interface. When the DB implementation is removed we can change the interface so that we could save some Get calls

Reply:
Sounds legit
Do we have an issue to follow up on thins?

Reply:
Opened: https://github.com/treeverse/lakeFS/issues/3695

Reply:
Thanks